db/buffered_insert.go

```diff
@@ -0,0 +1,218 @@
+func NewBufferedInserter(ctx context.Context, db Database, opts BufferedInsertOptions) (BufferedInserter, error) {
+	if err := opts.Validate(); err != nil {
+		return nil, errors.Wrap(err, "cannot construct buffered insert handler")
+	}
+
+	bi := &anserBufInsertsImpl{
+		docs:    make(chan interface{}),
+		err:     make(chan error),
+		flusher: make(chan chan error),
+		closer:  make(chan chan error),
+		db:      db,
+		opts:    opts,
+	}
+	ctx, bi.cancel = context.WithCancel(ctx)
+
+	go bi.start(ctx)
+
+	return bi, nil
+}
+
+func NewBufferedSessionInserter(ctx context.Context, s *mgo.Session, opts BufferedInsertOptions) (BufferedInserter, error) {
+	if opts.DB == "" {
+		return nil, errors.New("must specify a database name when constructing a buffered insert handler")
+	}
+
+	if s == nil {
+		return nil, errors.New("must specify a non-nil database session")
+	}
+
+	session := WrapSession(s)
+
+	return NewBufferedInserter(ctx, session.DB(opts.DB), opts)
+}
+
+func (bi *anserBufInsertsImpl) start(ctx context.Context) {
+	timer := time.NewTimer(bi.opts.Duration)
+	defer timer.Stop()
+	buf := []interface{}{}
+	catcher := grip.NewBasicCatcher()
+
+bufferLoop:
+	for {
+		select {
+		case <-ctx.Done():
+			if len(buf) > 0 {
+				catcher.Add(errors.Errorf("buffered insert has %d pending inserts", len(buf)))
+			}
+
+			bi.err <- catcher.Resolve()
+			break bufferLoop
+		case <-timer.C:
+			if len(buf) > 0 {
+				err := bi.db.C(bi.opts.Collection).Insert(buf...)
+				catcher.Add(err)
+				if err == nil {
+					buf = []interface{}{}
+				}
+			}
+
+			timer.Reset(bi.opts.Duration)
+		case doc := <-bi.docs:
+			buf = append(buf, doc)
+
+			if len(buf) >= bi.opts.Count {
+				err := bi.db.C(bi.opts.Collection).Insert(buf...)
+				catcher.Add(err)
+				if err == nil {
+					buf = []interface{}{}
+				}
+			}
+		case f := <-bi.flusher:
+			select {
+			case last := <-bi.docs:
+				buf = append(buf, last)
+			default:
+			}
+
+			if len(buf) > 0 {
+				err := bi.db.C(bi.opts.Collection).Insert(buf...)
+				catcher.Add(err)
+				if err == nil {
+					buf = []interface{}{}
+				}
+				f <- err
+				timer.Reset(bi.opts.Duration)
+			}
+
+			close(f)
+		case c := <-bi.closer:
+			close(bi.docs)
+			for last := range bi.docs {
+				buf = append(buf, last)
+			}
+
+			if len(buf) > 0 {
+				err := bi.db.C(bi.opts.Collection).Insert(buf...)
+				catcher.Add(err)
+			}
+			c <- catcher.Resolve()
+			close(c)
+			bi.cancel = nil
+			break bufferLoop
+		}
+	}
+
+	close(bi.err)
+
+}
+
+func (bi *anserBufInsertsImpl) Close() error {
+	if bi.cancel == nil {
+		return nil
+	}
+
+	res := make(chan error)
+	bi.closer <- res
+
+	bi.cancel()
+	bi.cancel = nil
+
+	return <-res
+}
+
+func (bi *anserBufInsertsImpl) Append(doc interface{}) error {
+	if doc == nil {
+		return errors.New("cannot insert a nil document")
+	}
+
+	bi.docs <- doc
+	return nil
+}
+
+func (bi *anserBufInsertsImpl) Flush() error {
+	res := make(chan error)
+	bi.flusher <- res
```

Comment:
You receive twice from an empty channel?

Reply:
no, you make the channel, send it off through the other channel and then wait for someone to send through it.